### PR TITLE
Add TrustZone support for Arm Cortex-M processors

### PR DESCRIPTION
**Pull request by nvt**
## Overview

This pull request introduces TrustZone support in Contiki-NG for Arm Cortex-M processors. TrustZone makes it possible to divide the system into a _secure world_ and a _normal world_. Hence, it can provide a Trusted Execution Environment that protects selected memory and hardware peripherals from the main system operating in the normal (unprivileged) world.

The main purpose of this pull request is to provide a foundation for TrustZone in Contiki-NG upon which more advanced functionality can be implemented.

The PR includes contributions from John Kanwar (initial prototype), Nicolas Tsiftes, Niclas Finne, Peter Jonsson, and Joakim Eriksson.

## Technical notes

The design is based on a combined firmware image that contains two Contiki-NG firmwares: one for the secure world and one for the normal world. To communicate between the two worlds, one can call a set of designated functions from the normal world to the secure world. Conversely, one can also call designated functions in the normal world from the secure world. A small API for this purpose is declared in `arch/cpu/arm/cortex-m/trustzone/tz-api.h`.

Currently, the only platform that is supported is the nRF5340, which provides TrustZone functionality in the application processor. A platform-specific example shows how one can build and program the device with secure world and normal world examples.

As future work, we consider extensions with support for more platforms, more API functions, and more extensive fault handling.

## File organization

There are three main commits as part of this pull request, each concerning distinct areas in the Contiki-NG directory structure.

1. The core TrustZone module is located in directory `arch/cpu/arm/cortex-m/trustzone`.

2. Platform-specific functionality for nRF5340 is placed in various files in `arch/cpu/nrf`.

3. The example software for TrustZone is located in `examples/platform-specific/nrf/trustzone`. Once more Contiki-NG platforms get TrustZone, this example may be moved.

## Testing

Enter `examples/platform-specific/nrf/trustzone` and follow the instructions in the README.md file. Note that the example works only with the nRF5340 development kit.

### CODE REVIEW

File: examples/platform-specific/nrf/trustzone/normal-world/Makefile [hunk 1]
```diff
@@ -0,0 +1,19 @@
+CONTIKI_PROJECT = normal-world-example
+CONTIKI = ../../../../..
```
Comment (pjonsson) [inline examples/platform-specific/nrf/trustzone/normal-world/Makefile:2]: No need to define `CONTIKI` nowadays.

File: examples/platform-specific/nrf/trustzone/secure-world/Makefile [hunk 2]
```diff
@@ -0,0 +1,18 @@
+CONTIKI_PROJECT = secure-world-example
+CONTIKI = ../../../../..
```
Comment (pjonsson) [inline examples/platform-specific/nrf/trustzone/secure-world/Makefile:2]: No need to define `CONTIKI` nowadays.

File: examples/platform-specific/nrf/trustzone/secure-world/secure-world-example.c [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  static struct etimer et;
+
+  PROCESS_BEGIN();
+
+  PROCESS_WAIT_EVENT_UNTIL(ev == trustzone_init_event);
+
+  LOG_INFO("Secure application ready to run\n");
+
+  etimer_set(&et, 60 * CLOCK_SECOND);
+  for(;;) {
```
Comment (pjonsson) [inline examples/platform-specific/nrf/trustzone/secure-world/secure-world-example.c:56]: Most other examples use `while(1)` instead of `for(;;)`.

File: examples/platform-specific/nrf/trustzone/normal-world/Makefile [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+all: $(CONTIKI_PROJECT)
+
+PLATFORMS_ONLY += nrf
+BOARDS_ONLY = nrf5340/dk/application
+
+TARGET ?= nrf
+BOARD ?= nrf5340/dk/application
+
+SECURE_WORLD_PATH ?= ../secure-world
+CMSE_IMPORT_LIB ?= CMSE-import-lib.o
```
Comment (pjonsson) [inline examples/platform-specific/nrf/trustzone/normal-world/Makefile:13]: Isn't this the default value for this variable?

File: examples/platform-specific/nrf/trustzone/normal-world/Makefile [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+TRUSTZONE_SECURE_BUILD = 0
+
+LDFLAGS += $(SECURE_WORLD_PATH)/build/nrf/nrf5340/dk/application/$(CMSE_IMPORT_LIB)
```
Comment (pjonsson) [inline examples/platform-specific/nrf/trustzone/normal-world/Makefile:17]: This line should be in `arch/cpu/arm/Makefile.arm` wrapped in `ifeq ($(TRUSTZONE_SECURE_BUILD), 0))`

File: arch/cpu/nrf/sys/clock-arch.c [hunk 6]
```diff
@@ -131,19 +143,20 @@ clock_time(void)
...
 }
 /*---------------------------------------------------------------------------*/
-void
+static void
 clock_update(void)
 {
   ticks++;
-  if(etimer_pending()) {
+  if(etimer_pending() &&
+     (etimer_next_expiration_time() - (clock_time_t)ticks - 1) > MAX_TICKS) {
```
Comment (pjonsson) [inline arch/cpu/nrf/sys/clock-arch.c:151]: This looks like some unrelated bug fix, does it belong in a separate PR?

File: arch/cpu/nrf/sys/clock-arch.c [hunk 7]
```diff
@@ -131,19 +143,20 @@ clock_time(void)
...
     etimer_request_poll();
   }
 }
 /*---------------------------------------------------------------------------*/
 unsigned long
 clock_seconds(void)
 {
-  return (unsigned long)ticks / CLOCK_CONF_SECOND;
+  return (unsigned long)(ticks / CLOCK_SECOND);
```
Comment (pjonsson) [inline arch/cpu/nrf/sys/clock-arch.c:159]: This looks like some unrelated bug fix, does it belong in a separate PR?

File: arch/cpu/arm/cortex-m/trustzone/normal/Makefile.normal [hunk 8]
```diff
@@ -0,0 +1,2 @@
+# Exclude the normal dbg-arch.c from the build
+CONTIKI_SOURCES_EXCLUDES += dbg-arch.c
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/normal/Makefile.normal:2]: Instead of first adding and then removing the file, wrap the place that adds this file with a `ifndef TRUSTZONE_SECURE_BUILD`.

File: arch/cpu/arm/cortex-m/trustzone/normal/tz-dbg-arch.c [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    dbg_pos = 0;
+  }
+
+  return c;
+}
+/*---------------------------------------------------------------------------*/
+unsigned int
+dbg_send_bytes(const unsigned char *s, unsigned int len)
+{
+  unsigned int i = 0;
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/normal/tz-dbg-arch.c:66]: I have just removed a number of copies of these functions, and #2650 is still in the PR queue. Put TrustZone ifdefs in the other file instead of making a copy of the file.

File: arch/cpu/arm/cortex-m/trustzone/normal/tz-normal.c [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+#include "contiki.h"
+#include "sys/platform.h"
+#include "trustzone/tz-api.h"
+
+/*---------------------------------------------------------------------------*/
+#include "sys/log.h"
+#define LOG_MODULE "TZNormalWorld"
+#define LOG_LEVEL LOG_LEVEL_INFO
+/*---------------------------------------------------------------------------*/
+static volatile bool is_poll_requested = false;
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/normal/tz-normal.c:51]: Isn't false already the default value for this variable?

File: arch/cpu/arm/cortex-m/trustzone/normal/tz-normal.c [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+  process_poll(&tz_normal_process);
+  return true;
+}
+/*---------------------------------------------------------------------------*/
+static void
+init_tz_api(void)
+{
+  struct tz_api tz_api;
+
+  memset(&tz_api, 0, sizeof(tz_api));
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/normal/tz-normal.c:68]: It's shorter with a `= {0}` in the declaration of `tz_api` instead.

File: arch/cpu/arm/cortex-m/trustzone/tz-api.c [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+static volatile int is_poll_requested;
+
+/*---------------------------------------------------------------------------*/
+#include "sys/log.h"
+#define LOG_MODULE "TZAPI"
+#define LOG_LEVEL LOG_LEVEL_INFO
+/*---------------------------------------------------------------------------*/
+process_event_t trustzone_init_event;
+/*---------------------------------------------------------------------------*/
+CC_TRUSTZONE_SECURE_CALL
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/tz-api.c:58]: Shouldn't this be on the same line as `bool`?

File: arch/cpu/arm/cortex-m/trustzone/tz-api.c [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+  }
+
+  trustzone_init_event = process_alloc_event();
+
+  apip = cmse_check_address_range(apip, sizeof(*apip), CMSE_NONSECURE);
+  if(apip == NULL) {
+    return false;
+  }
+
+  if(apip->request_poll == NULL) {
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/tz-api.c:73]: Combine with the previous so all apip null checks are in the same if-statement.

File: arch/cpu/arm/cortex-m/trustzone/tz-api.c [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+    watchdog_periodic();
+  }
+
+  is_poll_running = false;
+  is_poll_requested = 0;
+
+  return process_nevents() > 0;
+}
+/*---------------------------------------------------------------------------*/
+CC_TRUSTZONE_SECURE_CALL
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/tz-api.c:123]: Shouldn't this be on the same line as void?

File: arch/cpu/arm/cortex-m/trustzone/tz-api.c [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+void
+tz_api_println(const char *text)
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/tz-api.c:125]: Can't this function have a printf-like signature, so the caller doesn't have to call sprintf?

File: arch/cpu/arm/cortex-m/trustzone/tz-api.c [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+#include "sys/autostart.h"
+#include "tz-api.h"
+
+#include <arm_cmse.h>
+#include <stdarg.h>
+#include <string.h>
+
+static struct tz_api tz_api;
+static bool initialized;
+static volatile int is_poll_requested;
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/tz-api.c:49]: This is a counting variable, but the name `is_` implies it is some kind of boolean.

File: arch/cpu/arm/cortex-m/trustzone/tz-api.h [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+ * @{
+ */
+
+#ifndef TZ_API_H
+#define TZ_API_H
+
+#include <stdbool.h>
+#include <stdlib.h>
+
+#ifdef NRF_TRUSTZONE_SECURE
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/tz-api.h:77]: Doesn't this need to be wrapped around each macro for Doxygen output to look reasonable?

Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/tz-api.h:77]: Doesn't this need to be wrapped around each macro for Doxygen output to look reasonable?

File: arch/cpu/arm/cortex-m/trustzone/tz-api.c (hunk 15 above)
Comment (nvt) [reply arch/cpu/arm/cortex-m/trustzone/tz-api.c:118]: This is an internal API functionthat is always supplied with a buffered string.

File: arch/cpu/nrf/dev/watchdog-arch.c [hunk 18]
```diff
@@ -63,12 +68,15 @@ static uint8_t wdt_initialized = 0;
 static void
 wdt_event_handler(void)
 {
```
Comment (pjonsson) [inline arch/cpu/nrf/dev/watchdog-arch.c:70]: This stops turning off leds on non-TrustZone platforms. Is that intentional?

File: arch/cpu/arm/cortex-m/trustzone/tz-fault.c [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+static void
+print_sfsr(uint32_t sfsr)
+{
+  if(sfsr & SAU_SFSR_LSERR_Msk) {
+    LOG_WARN_(" LSERR");
+  }
+  if(sfsr & SAU_SFSR_SFARVALID_Msk) {
+    LOG_WARN_(" SFARVALID");
+  }
+  if(sfsr & SAU_SFSR_LSERR_Msk) {
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/tz-fault.c:70]: Is this different from line 64?

File: arch/cpu/arm/cortex-m/trustzone/tz-fault.c [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+    LOG_WARN_(" INVTRAN");
+  }
+  if(sfsr & SAU_SFSR_AUVIOL_Msk) {
+    LOG_WARN_(" AUVIOL");
+  }
+  if(sfsr & SAU_SFSR_INVER_Msk) {
+    LOG_WARN_(" INVER");
+  }
+  if(sfsr & SAU_SFSR_INVIS_Msk) {
+    LOG_WARN_(" LSERR");
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/tz-fault.c:86]: Should INVIS print LSERR?

File: arch/cpu/arm/cortex-m/trustzone/tz-secure.c [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+    return NULL;
+  }
+
+  enum tfm_plat_err_t tfm_err = nvic_interrupt_target_state_cfg();
+  if(tfm_err != TFM_PLAT_ERR_SUCCESS) {
+    LOG_DBG("Interrupt state: 0x%x\n", tfm_err);
+  }
+
+  unsigned int *vtor_ns = (unsigned int *)NS_CODE_START;
+  LOG_DBG("NS image at 0x%x\n", (unsigned int)vtor_ns);
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/tz-secure.c:96]: The cast for pointers to integers should be `(u)intptr_t`.

File: arch/cpu/arm/cortex-m/trustzone/tz-secure.c [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+  LOG_DBG("Secure gateway region: %p - %p\n", &__sg_start, &__sg_end);
+  LOG_DBG("NSC size %p\n", &__nsc_size);
+  spu_regions_flash_config_non_secure_callable((uint32_t)&__sg_start,
+					       (uint32_t)&__sg_end - 1);
+
+  LOG_DBG("__ARM_FEATURE_CMSE = %d\n", __ARM_FEATURE_CMSE);
+
+  /* Configure non-secure stack */
+  tz_nonsecure_setup_conf_t spm_ns_conf = {
+      .vtor_ns = (unsigned int)vtor_ns,
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/tz-secure.c:113]: Same here, the cast for pointers to integers should be `(u)intptr_t`.

File: arch/cpu/arm/cortex-m/trustzone/Makefile.trustzone [hunk 23]
```diff
@@ -0,0 +1,4 @@
+# Provide a library of secure functions to the normal world.
+LDFLAGS += -Xlinker --cmse-implib
+LDFLAGS += -Xlinker --out-implib=$(BUILD_DIR_BOARD)/$(CMSE_IMPORT_LIB)
+LDFLAGS += -Xlinker --sort-section=alignment
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/Makefile.trustzone:4]: Doesnt Makefile.cortex-m:35 already contain `--sort-section=alignment`?

File: arch/cpu/arm/cortex-m/Makefile.cortex-m [hunk 24]
```diff
@@ -15,6 +15,22 @@ CONTIKI_ARM_DIRS += cortex-m
 ### Build syscalls for newlib
 MODULES += os/lib/newlib
 
+ifdef TRUSTZONE_SECURE_BUILD
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/Makefile.cortex-m:18]: This ifdef can be  removed, nothing uses `CMSE_IMPORT_LIB` unless building for TrustZone.

File: arch/cpu/arm/cortex-m/trustzone/Makefile.trustzone [hunk 25]
```diff
@@ -0,0 +1,4 @@
+# Provide a library of secure functions to the normal world.
+LDFLAGS += -Xlinker --cmse-implib
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/Makefile.trustzone:2]: If you put the first two lines of LDFLAGS under the `ifeq ($(TRUSTZONE_SECURE_BUILD), 1)`  block in Makefile.cortex-m instead, this Makefile can be removed completely.

File: arch/cpu/arm/cortex-m/trustzone/normal/Makefile.normal [hunk 26]
```diff
@@ -0,0 +1,8 @@
...
+
+ifndef TRUSTZONE_SECURE_WORLD_PATH
+  $(error TRUSTZONE_SECURE_WORLD_PATH must be set.)
+endif
+
+LDFLAGS += $(TRUSTZONE_SECURE_WORLD_PATH)/$(BUILD_DIR_BOARD)/$(CMSE_IMPORT_LIB)
```
Comment (pjonsson) [inline arch/cpu/arm/cortex-m/trustzone/normal/Makefile.normal:8]: If you put the last five lines under the `else ifeq ($(TRUSTZONE_SECURE_BUILD), 0)` block in Makefile.cortex-m instead, this Makefile can be removed completely.

File: arch/cpu/nrf/nrf5340/tz-target-cfg.c [hunk 27]
```diff
@@ -0,0 +1,410 @@
...
+   * Allow SPU to have precedence over (non-existing) ARMv8-M SAU.
+   */
+  TZ_SAU_Disable();
+  SAU->CTRL |= SAU_CTRL_ALLNS_Msk;
+}
+/******************************************************************************/
+enum tfm_plat_err_t
+spu_periph_init_cfg(void)
+{
+  /* LOG_DBG("PERIPHERAL START \n"); */
```
Comment (pjonsson) [inline arch/cpu/nrf/nrf5340/tz-target-cfg.c:129]: From upstream, or leftover debug code?

File: arch/cpu/nrf/nrf5340/tz-target-cfg.c [hunk 28]
```diff
@@ -0,0 +1,410 @@
...
+  spu_peripheral_config_non_secure((uint32_t)NRF_SPIM3, false);
+
+  NVIC_DisableIRQ(NRFX_IRQ_NUMBER_GET(NRF_SAADC));
+  spu_peripheral_config_non_secure((uint32_t)NRF_SAADC, false);
+
+  NVIC_DisableIRQ(NRFX_IRQ_NUMBER_GET(NRF_TIMER0));
+  spu_peripheral_config_non_secure((uint32_t)NRF_TIMER0, false);
+
+  /* NVIC_DisableIRQ(NRFX_IRQ_NUMBER_GET(NRF_TIMER1)); */
+  /* spu_peripheral_config_non_secure((uint32_t)NRF_TIMER1, false); */
```
Comment (pjonsson) [inline arch/cpu/nrf/nrf5340/tz-target-cfg.c:168]: If these lines are from upstream, future merges will be easier if it is instead disabled with:
```
#if 0
  <upstream code>
#endif
```

File: arch/cpu/nrf/nrf5340/tz-target-cfg.c [hunk 29]
```diff
@@ -0,0 +1,410 @@
...
+
+  NVIC_DisableIRQ(NRFX_IRQ_NUMBER_GET(NRF_TIMER2));
+  spu_peripheral_config_non_secure((uint32_t)NRF_TIMER2, false);
+
+  /* Before commented. */
```
Comment (pjonsson) [inline arch/cpu/nrf/nrf5340/tz-target-cfg.c:173]: Upstream comment, or was the next statements commented out in the upstream source?

File: arch/cpu/nrf/nrf5340/tz-target-cfg.c [hunk 30]
```diff
@@ -0,0 +1,410 @@
...
+  nrf_gpio_pin_mcu_select(PIN_XL1, NRF_GPIO_PIN_MCUSEL_PERIPHERAL);
+  nrf_gpio_pin_mcu_select(PIN_XL2, NRF_GPIO_PIN_MCUSEL_PERIPHERAL);
+
+  /*
+   * Enable the instruction and data cache (this can be done only from secure
+   * code; that's why it is placed here).
+   */
+  NRF_CACHE->ENABLE = CACHE_ENABLE_ENABLE_Enabled;
+
+  /* LOG_DBG("PERIPHERAL STOP \n"); */
```
Comment (pjonsson) [inline arch/cpu/nrf/nrf5340/tz-target-cfg.c:325]: Leftover debug printouts?

File: arch/cpu/nrf/sys/clock-arch.c (hunk 6 above)
Comment (nvt) [reply arch/cpu/nrf/sys/clock-arch.c]: Yes, the clock changes will become a separate PR.

File: arch/cpu/arm/Makefile.arm [hunk 31]
```diff
@@ -24,7 +24,7 @@ CFLAGS += -std=c99
 CFLAGS += -ffunction-sections -fdata-sections -fno-strict-aliasing
 CFLAGS += -fshort-enums -fomit-frame-pointer
 ifeq ($(TRUSTZONE_SECURE_BUILD),1)
-  CFLAGS += -mcmse
+  CFLAGS += -mcmse --specs=nosys.specs
```
Comment (pjonsson) [inline arch/cpu/arm/Makefile.arm:27]: Clang does not have `--specs`, so put it in a separate line:
```
ifeq ($(CLANG),0)
  CFLAGS += --specs=nosys.specs
endif
```

File: arch/cpu/nrf/Makefile.nrf [hunk 32]
```diff
@@ -8,8 +8,8 @@ CONTIKI_CPU_DIRS += net
 ### CPU-dependent source files
 CONTIKI_CPU_SOURCEFILES += rtimer-arch.c
 CONTIKI_CPU_SOURCEFILES += watchdog-arch.c
-CONTIKI_CPU_SOURCEFILES += clock-arch.c
 CONTIKI_CPU_SOURCEFILES += dbg-arch.c
+CONTIKI_CPU_SOURCEFILES += clock-arch.c
```
Comment (pjonsson) [inline arch/cpu/nrf/Makefile.nrf:12]: Accidental change?

**Review by nfi — APPROVED** [review]

**Review by joakimeriksson — APPROVED** [review]
